opal/browser/utils.rb

```diff
@@ -3,6 +3,41 @@ module Browser
   Position = Struct.new(:x, :y)
 end
 
+module Native
+  module Wrapper
+    # {Native::Wrapper::Singleton} is a special case of {Native::Wrapper}.
+    #
+    # What this module does is it makes sure that your Ruby objects
+    # are mapped 1:1 to your Javascript objects. So that for instance
+    # your `$document.at_css('body')` is always the same Ruby object.
+    #
+    # You can only use it if your final `.new` is of the signature
+    # `.new(native)` and your native (probably DOM) object persists and
+    # doesn't mind arbitrary properties.
+    #
+    # The rule of thumb is: if it does overload `#initialize`'s signature
+    # and not `.new`'s - it won't work. Use Native::Wrapper in this case.
```

Comment:
It all makes sense, but it needs a couple of adjustments.

1. It's better to stay within the `Browser` namespace and have a local, specific wrapper (that also helps in not polluting the constants space of every class that already includes `Native::Wrapper`)
2. As of the name I think something like `Browser::CachedNativeWrapper`, but possibly something even better (please look for variations), as you said `Singleton` is generally interpreted with another meaning, although it makes sense once you reason about it.
3. I would use something like `$$opal_browser_native` or `$$native` instead of `__opal_singleton__` because that wouldn't show up in instance variables, and it's the common practice in opal.

Reply:
I think this module isn't necessarily opal-browser only and maybe someday it would be wise to move it to some more general repo, like opal. After all, it's not tied to DOM, but for now it's most needed by DOM. I want to keep its name as similar to Native::Wrapper as possible (because some classes include one, others include the other), so maybe Native::CachedWrapper and $$opal_native_cached?  If staying inside the namespace is more important, then Browser::NativeCachedWrapper will be the choice I will take for this pull request.

Reply:
By the way, the module has another ClassMethods module inside. I think it's the Ruby convention to do extends this way, but it will pollute the namespace anyway.

Reply:
So, the thing is, this looks like something that should end up in opal's stdlib. If by any chance the final version has a different api that opal-browser's version it will give us some trouble in reconciling the two, that's why I'd like to keep it in a namespace. The name you proposed and the desire to keep the names similar are 👍 so, what about `Browser::Native::CachedWrapper`?

WRT `ClassMethods` I'm not sure where it's coming from, but the stdlib module seems clean: https://github.com/opal/opal/blob/3eae98d70045b3daa56ca6656e3331c92ed18e57/stdlib/native.rb#L189-L212

Reply:
Good idea about the naming, I'm preparing the patch right now.

I override `new` which is a class method, so I need to extend it but simultaneously I want to include Native::Wrapper. The established way to do is to create a submodule and extend the class with this submodule inside self.included. Now I see that the submodule gets added as a constant as well. I moved it a level up and renamed it to `CachedWrapperClassMethods`. In a moment I will produce the commit.

Reply:
Oh wait, that's not really a good idea. This caused a lot of tests to fail with

`undefined method `convert' for Browser::Native`

And a few related errors like Browser::Native::Array etc. I will use the Browser::NativeCachedWrapper name.

Reply:
constant lookup is always fun 😄